lib/rule_runner.rb

```diff
@@ -0,0 +1,52 @@
+class RuleRunner
+  def initialize(rule:, callback:)
+    @rule = rule
+    @callback = callback
+  end
+
+  def perform
+    rule_event.add_detail(:info, "Event received")
+
+    rule.filters.each.with_index do |filter, index|
+      if FilterValidator.new(filter).valid?(callback)
+        rule_event.add_detail(:info, "Filter ##{index + 1}: Met")
+      else
+        rule_event.add_detail(:info, "Filter ##{index + 1}: Unmet, event dropped")
+        return
+      end
+    end
```

Comment:
Do we assume filters are always ordered, in the same way, or should we capture the return value of all filters, log it, and then return if one of them is invalid?

Reply:
So 2 questions here:

> Do we assume filters are always ordered

Let's throw in an `order(:id)` to guarantee the behaviour.

For now I settled with "Filter #X" but we should revisit this to see if it can be made better visually. 

> Should we capture the return value of all filters, log it, and then return if one of them is invalid?

Why not. I can do this in this PR if it is what you would like.

Reply:
Updated to run all filters and log the result of each (will still stop the process if any fail)